[export] Add lifted constant obj to input

angelayi opened the pull request:
Test Plan: wip

Differential Revision: D52556070

suo reviewed (changes requested):
mostly looks good! Had some questions inline about the implementation

angelayi commented on torch/_export/serde/serialize.py:
```diff
@@ -1292,6 +1318,21 @@ def deserialize_signature(self, sig: GraphSignature) -> ep.ExportGraphSignature:
             output_specs=[self.deserialize_output_spec(o) for o in sig.output_specs]
         )
 
+    def update_custom_obj_node_meta(self, graph, signature, constants):
```
This is assuming that our `val` is the custom obj itself, if we serialize to `serialized_name_to_meta`, then the serialized custom object will live in both `serialized_name_to_meta`, which is attached to the serialized exported program, and in `constants`, which is saved at a different location from the exported program. I was thinking it's probably not good to save one thing in two places, and especially if it's a custom object blob..?

suo replied:
`serialized_name_to_meta` is a field on `GraphModuleDeserializer` and isn't persisted to the final `ExportedProgram`, as far as I can tell. It's just bookkeeping to map an IR name to `meta['val']`, which seems like what we're doing here. So tracking it there should not affect the serialized format.

I think the more fundamental issue is that we don't have a serialized table mapping names to custom objs in `schema.Graph` like we do for every other dynamic value type (tensors, symints, symbools). We are representing it implicitly through placeholder arguments/inputs to nodes. 

The asymmetry of how we handle these nodes is a good signal that there's a missing piece in our design. The missing piece is that we don't have a metadata object for ScriptObjects: `Tensor : TensorMeta :: ScriptObject : ???`.

If we had such a piece (call it `ScriptObjectMeta`), the right fix would be:
1. `schema.Graph` gets a `custom_object_values: Dict[str, ScriptObjectMeta]` field, analogous to `tensor_values`. 
2. We deserialize that and populate `serialized_name_to_meta` like everything else in `GraphModuleDeserializer.deserialize_graph`.
3. `meta["val"]` should contain `ScriptObjectMeta`, not the `ScriptObject` itself. 

Okay, so what should `ScriptObjectMeta` look like? If we were being faithful to the `Tensor : TensorMeta` analogy, it should be something that we can deserialized into a `FakeScriptObject` that will work when invoking the node in `FakeMode`. So that's probably the right thing.

That's out of scope for this diff. But we should try to design things with a structure **as close as possible** to the right thing, so that when we go fix it later it will be easier (and we don't create weird divergences in the meantime). 

My suggestion for a temporary `ScriptObjectMeta` would be a dataclass that looks like:
```
@dataclass
class ScriptObjectMeta:
  name: str  # key into constants table to retrieve the real ScriptObject
  typename: str  # fully qualified type of the ScriptObject. 
```

What do you think?

angelayi replied:
From your comment I realized I was implementing it stupidly, so I updated the code, but not exactly what you said. Let me know what you think! Basically I just populated `serialized_name_to_meta` based on the signature, since the signature stores the custom obj's node name and fqn to the `constants` dictionary.

angelayi replied:
Based on what we discussed, I added a `ScriptObjectMeta`, which stores just the name of the constant in the constants table.

angelayi commented on torch/_export/serde/serialize.py:
```diff
@@ -1410,6 +1453,9 @@ def deserialize_optional_tensor_args(a):
             else:
                 raise SerializeError(f"Unhandled argument {inp}")
         elif typ_ == "as_custom_obj":
+            if inp.as_custom_obj.name in self.serialized_name_to_node:
```
This is for the current flow which nvidia TRT uses, which just inlines the custom obj. Once we get tracing to work, they can just retrace and it'll lift the custom obj.

angelayi commented on torch/export/graph_signature.py:
```diff
@@ -264,6 +276,17 @@ def lifted_tensor_constants(self) -> Collection[str]:
+    # A list of lifted custom objects
+    @property
+    def lifted_custom_objs(self) -> Collection[str]:
+        # TODO Make this tuple.
```
I'm not sure 😅 I think @zhxchen17 added the TODOs in the other fields so I just followed suit and was going to let him fix it.

angelayi commented on torch/export/exported_program.py:
```diff
@@ -231,6 +232,11 @@ def dialect(self) -> str:
+    @property
+    @compatibility(is_backward_compatible=False)
+    def constants(self):
+        return self._tensor_constants
```
I'll do a renaming in most of the export code, but won't get rid of the attribute here (so I don't break executorch ci 🙈)

angelayi committed:
[export] Add lifted constant obj to input (#116985)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/116985

This is in preparation for tracing support for custom class holder objects. We expect them to initially appear in the graph as getattr nodes. This diff will then lift the getattr nodes as inputs to the graph with the graph signature being `InputKind.CUSTOM_OBJ`, and adds serialization support.

Test Plan: Added test

Reviewed By: suo

Differential Revision: D52556070

fbshipit-source-id: be6743c5a94b115052274b7afbe54711f76d5b80
